Adding and Refactoring of Tests

localthomas opened the pull request:
Short overview: there are two test patterns of 4×4 pixel blocks available: `GRAY` and `COLOUR`.
`GRAY` is a gray-scale checkerboard starting with 0xFF in the top-left. On top of that, the four middle pixels are set to 0x7F or 0x55.
`COLOUR` is a colour test-pattern (RGB) with the first row in one colour, the second in another and the third and last row in a third colour. The colours are (0xFF, 0x96, 0x4A), (0xFF, 0x78, 0x34), and (0xFF, 0x69, 0x29). Originating from #2.

The following unit tests were added (all encoding and decoding):
* BC2 `GRAY`
* BC3 `GRAY`
* BC3 `COLOUR`
* BC4 `GRAY` (with a small bugfix for expanding the single channel into RGB)
* BC5 `GRAY` (variant with red being `GRAY` and green being the inverse of `GRAY`)

---

Additionally, the tests were refactored and a new module `test_data` introduced. This module holds the static data required for the unit tests. See the comments in the module for more information.

Let me know what you think, especially about the new module.

localthomas committed:
added BC2 unit tests with gray pattern

localthomas committed:
renamed static test data to common scheme

localthomas committed:
extracted static test data for unit tests into new module

localthomas committed:
simplified compression and decompression test cases

localthomas committed:
added BC3 unit tests

localthomas committed:
added BC4 unit tests and fixed splatting error

localthomas committed:
added BC5 unit tests

jansol commented:
Sorry about the delay, had a bit of an emergency situation eating away my time for this crate. I'll check this out a bit later tonight.

jansol commented on squish/src/test_data.rs:
```diff
@@ -0,0 +1,202 @@
+
+/// Appends a list of 16 alpha values to the RGB values.
+/// I.e. each RGB is extended to RGBA with the alpha value from the list.
+const fn add_custom_alpha_to_rgb(
```
```suggestion
const fn add_alpha_to_rgb(
```

jansol commented on squish/src/test_data.rs:
```diff
@@ -0,0 +1,202 @@
+}
+
+/// Adds to the RGB data the alpha value and returns the resulting RGBA data.
+const fn add_alpha_to_rgb(input: &[u8; 4 * 4 * 3], alpha: u8) -> [u8; 4 * 4 * 4] {
```
This is redundant since you can just pass `&[alpha; 4*4]` to the generic fn. The dedicated fn for this case doesn't make it much shorter to write anyway.

jansol commented on squish/src/test_data.rs:
```diff
@@ -0,0 +1,202 @@
+
+/// Creates a linear alpha ramp starting with the first pixel at 0x00 and ending at 0xFF.
+/// The step size is therefore 0x11.
+const fn add_alpha_ramp_to_rgb(input: &[u8; 4 * 4 * 3]) -> [u8; 4 * 4 * 4] {
```
Ditto, add something like
```
const LINEAR_RAMP: [u8; 4*4] = [
  0x00, 0x11, 0x22, 0x33,
  0x44, 0x55, 0x66, 0x77,
  0x88, 0x99, 0xAA, 0xBB,
  0xCC, 0xDD, 0xEE, 0xFF
];
```
and just use that with the generic implementation instead of this fn.

jansol reviewed (changes requested):
Looks good overall, but could be a bit more DRY.

localthomas committed:
removed two const fn for generating test data

Two const fn for adding alpha values to RGBA data arrays were removed to reduce code duplication.
Call-sites now use array literals or a constant linear ramp value.

localthomas commented:
I implemented all suggestions from the review. Removing two of the three nearly redundant functions was a good idea.

---
> Sorry about the delay, had a bit of an emergency situation eating away my time for this crate. I'll check this out a bit later tonight.

No worries. I do the work in my spare time and fully respect that personal time comes first. There are no deadlines for these changes after all.
